### ISSUE TITLE
Data Queries with 2 Associated Graph Templates

### ISSUE THREAD

**Issue by kruegerj** [Cacti/cacti#2364]
I've a Data Queries self build with 2 Graph Templates assigned. Now, if i go to my Device and create these two Graphs only the first rra is created. 

Data Query:
![image](https://user-images.githubusercontent.com/8223375/52051116-de61ad00-2552-11e9-859b-f26c4c385b62.png)

0 Transmitted Traffic:
![image](https://user-images.githubusercontent.com/8223375/52051542-2df4a880-2554-11e9-8498-528e2060408d.png)

1 Dropped Traffic:
![image](https://user-images.githubusercontent.com/8223375/52051571-3ea51e80-2554-11e9-9a1f-8db7ac2f7b22.png)

Only if i check Data Source Active (Check this checkbox if you wish to allow the user to override the value on the right during Data Source creation.) and use the second click on creation, the 2nd rrd is created.
![image](https://user-images.githubusercontent.com/8223375/52051081-c68a2900-2552-11e9-8e3f-0e683baaa6d4.png)

Only one Data Source and two Graphs:
![image](https://user-images.githubusercontent.com/8223375/52051353-a6a73500-2553-11e9-8980-8709594f3714.png)

With workaround:
![image](https://user-images.githubusercontent.com/8223375/52051492-0bfb2600-2554-11e9-8975-f9d167c8de68.png)

**Comment by cigamit** [comment]
Well, if they are both collecting the same data, Cacti is now designed to not create the second data source.  It's normal behavior.  However, I'm thinking that you may have another problem.  I'm not not sure what it is yet.

**Comment by kruegerj** [comment]
Maybe that helps:
![image](https://user-images.githubusercontent.com/8223375/52070317-d882c080-2580-11e9-80f6-7bc4a7f7b641.png)
![image](https://user-images.githubusercontent.com/8223375/52070355-ed5f5400-2580-11e9-8b42-de01fbec1197.png)
you see, that both collecting different data.

**Comment by ian351c** [comment]
I think this might have something to do with issue #948 somehow. I have seen similar behavior. If I create a data query and associate it with a graph template that does not use all the data source items in the data source template associated with the graph template, the poller will only create an RRD and fill it with the data for the data source items used in that first graph template. Cacti does not appear to look at the other graph templates associated with a data query once that first graph template has been associated, at least as far as figuring out what data needs to be polled for and saved into RRDs.

I work around this by creating a graph template that uses all the data source items and make sure it is the first graph template associated with the data query.

**Comment by cigamit** [comment]
Well, I know where in the code it does this.  Generally reusing a data template for something completely different is not a great idea, as it's a corner case.  Can either of you dump your database and send it to developers at cacti dot net?

**Comment by ian351c** [comment]
I'm not comfortable sending my "production" database, but if I have time I'll replicate the issue in a fresh install. In the meantime, if you could expand on your "not a great idea" sentiment, that might be helpful. Are you saying that it's not a good idea to have one "big data source template" that defines a bunch of different data source items? For instance, in the stock "Interface" templates that come with Cacti, the In/Out Bits and In/Out Errors/Discarded Packets graph templates use separate data templates (Interface - Traffic and Interface - Errors/Discards) even though they're all in the same data query. Are you saying that if you had one big data source template that defined all 6 of those data source items (traffic_in, traffic_out, errors_on, errors_out, discards_in, and discards_out) and the two graph templates referenced that single data source template, that would be a "bad idea" and wouldn't work? That's fine if that's how 1.2.x and forward is going to be (and I understand the approach), but it would be good to note that this will break things that used to work. For instance, the template I first ran into this with is the Cacti NUT Template (https://forums.cacti.net/about19250.html). This template has a single data source template with six data source items along with three graph templates which reference different data source items within the data template. In my old 0.8.x Cacti installation, using this template results in 3 RRDs (one per graph template) that are technically "duplicate" RRDs since they all have the same structure (all the data source items are defined in each RRD). However, each RRD only has data in the particular data source items used for the associated graph and the poller knows what goes where. In a fresh install of 1.2, creating the first graph based on the template still results in an RRD with all the data source items defined, but the poller will only fetch data for the data source items referenced in the first graph template associated with the data query. Creating additional graphs based on the other graph templates results in those graphs being empty because they refer the the first RRD that was created, but the poller is only fetching data source items referenced in the first graph template associated with the data template. IMHO, the fix for this is either to fetch all the data source items defined in a data source by default, or to create logic that goes back and tells the poller that there are new data source items that need to be fetched for an RRD whenever a new graph is created.

TL;DR: if a "one to one" approach to graph templates and data templates is a best practice, then that's fine. (BTW: is this documented anywhere? I'll admit that I've never read the docs all the way through; I've always just searched for what I needed and used HOWTOs where possible). Just be aware that enforcing this best practice will break things that used to work.

**Comment by cigamit** [comment]
I can simulate with the 32bit vs. 64bit Data Queries Templates.  It's just nicer to have a second case.  So, you have to give me a few days as there is lot's do do, and not enough time.

**Comment by cigamit** [comment]
You are right that creating multiple Data Templates will correct the problem, but this issue I think affects the Interface graphs too.  So, I'll look into it.

**Comment by tertius1** [comment]
This happens with a stock data source template/graph template as well: NET-SNMP - Device I/O
This data source template has 7 data source items.
Associated are 3 graphs "NET-SNMP - Device I/O - Bytes Read/Written", "NET-SNMP - Device I/O - Bytes Load Averages", "NET-SNMP - Device I/O - Reads/Writes".
Each of the graphs graph 3 (Load Averages) or 2 (the other ones) of the 7 data source items.

To reproduce the issue (tested on a fresh cacti 1.2.1 install):
- Management->Devices->(some NET-SNMP device)->Create Graphs for this device
- navigate to the Data Query [Net-SNMP - Get Device I/O] section
- check a device, for example sda, and add each of the three available "NET-SNMP - Device I/O - xxxx" graphs.
-> 1 data source and 3 graphs will be created. The 3 graphs all reference the same *.rrd file. But data is collected only for the graph that was created first. The other data source items are not polled. There is no entry in the poller cache for these.

In general, the behavior of creating just one data source is perfectly fine, because it makes no sense to create 3 *.rrd files where in the 1st the 1st two data sources are filled only, in the 2nd the 3rd and 4th data sources are filled and in the last only items 5 to 7 are filled. It's a waste, and worse, the data source items are all offered at manual graph creation with identical names, where you don't know which is the correct one.

The problem is there are no polling requests in the poller cache being created for all items but the ones used by the first graph created.

There is something missing, it is the custom data of the data source where (for Device I/O) Index type, Index Value and Output Type ID is used to construct the poller cache entry. This index data is not saved anywhere for the additional graphs, so cacti isn't able to create the corresponding poller cache items.

Proposal: don't use "output type id" with these kind of data sources - use only "index value" and poll all existing output type id's. This way, the whole *.rrd file is filled; all data source items. Caveat: if only part of it is used, the additional data is polled for nothing.

**Comment by cigamit** [comment]
Understood.  There is something currently not working as expected.  The query that shows uniqueness appears to work as desired, but there is some logic flaw that allows the Data Source to be or not to be duplicated.  So, we will keep this open to address and mark it critical/blocking.  But we also need to get another release done as soon as this weekend.  Hopefully, it will make it into that release.

**Comment by cigamit** [comment]
@kruegerj, this should be resolved here.  https://github.com/Cacti/cacti/commit/4f3383ae7e66ad7f38d3d47d53a52982f5908801. Please test ASAP.

**Comment by kruegerj** [comment]
i'm sorry, but - with complete new fetch from github - no changes for me. 
There is only one DS. The second DS is missing.
![image](https://user-images.githubusercontent.com/8223375/52527425-b7c91200-2cc8-11e9-83de-ee36b7f7542f.png)
![image](https://user-images.githubusercontent.com/8223375/52527445-e5ae5680-2cc8-11e9-9c21-e3ff93f74c66.png)

And #2362 is also not fixed for me

**Comment by cigamit** [comment]
Okay, I just added tracing for this.  Goto Console > Configuration > Settings > General and enable Data Source Tracing.  Then create the Graphs for this issue and the Graphs for the rrd_maximum.  Post the trace output that is in your cacti.log.  Just search for DSTRACE.

**Comment by cigamit** [comment]
Please do this ASAP, but to be honest, I'm thinking that this may not be supportable.  That is to re-use the same internal data source name for non-compatible data.  I'm just not sure how we block it.  Let's do the tracing, and that will give me something to think about.

**Comment by cigamit** [comment]
So, I have implemented a workaround for now that allows the Traffic graphs to not duplicate graphs, while other Data Queries will always duplicate.  It's not pretty, but for now it will have to do until such time as we have a setting at the data query level to avoid Data Source duplication, and a check to verify that this setting will be acceptable.

**Comment by cigamit** [comment]
Any update guys?

**Comment by netniV** [comment]
@kruegerj @tertius1 @ian351c 

We are hoping to create a release this weekend, so any feedback on this would be useful.

**Comment by kruegerj** [comment]
I'm in vacation until 04.03. so i can't test at this time

**Comment by cigamit** [comment]
Closing.  1.2.2 is released.  Test and if still and issue, open a new bug referencing this one.
